mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp

```diff
@@ -369,6 +370,32 @@ LogicalResult spirv::Deserializer::processMemberName(ArrayRef<uint32_t> words) {
   return success();
 }
 
+void spirv::Deserializer::setArgAttrs(uint32_t argID) {
+  if (!decorations.contains(argID)) {
+    argAttrs.push_back(DictionaryAttr::get(context, {}));
```

Comment:
Hmm, this assumes we always process `argID` in increasing order, which is the case for now. But not guaranteed. It's better to use `resize` if `argID >= argAttrs.size()`.

Also as it's implemented right now, we have `argAttrs` in the class directly. So this will just continuously grow if we process more than one functions. So it would be nice to add one more test in the serialization tests, where we have at least two functions, and each function has more than one arguments:

```
spirv.func @fn1(%arg0: i32, %arg1: !spirv.ptr<..> {spirv.decoration = ...})
spirv.func @fn2(%arg0: !spirv.ptr<..> {spirv.decoration = ...}, %arg1: !spirv.ptr<..> {spirv.decoration = ...})
```

Reply:
> Also as it's implemented right now, we have argAttrs in the class directly. 

I move `argAttrs` into the `processFunction` as local variable.

> Hmm, this assumes we always process argID in increasing order, which is the case for now. But not guaranteed. It's better to use resize if argID >= argAttrs.size().

I understand that `argID` represents the ID for each variable's mapping, and it doesn't compare the length of `argAttrs`, correct? So my assumption is that the length of argAttrs matches the length of `getNumInputs`. In that case, why not reserve it as follows?

```cpp
LogicalResult
spirv::Deserializer::processFunction(ArrayRef<uint32_t> operands) {
  ...

  SmallVector<Attribute> argAttrs;
  argAttrs.reserve(functionType.getNumInputs());

  ...
```

Reply:
Yes `reserve` is helpful to allocate the space in one go. Though it does not initialize the elements--you'd still need to either `push_back` for all elements or call `resize` which initalizes elements along the way.